darwin: remove `nix search` warning spam

keysmashes opened the pull request:
Related to #407398.

This is a hacky mess but it does help: now `nix search` emits no Darwin-related warnings.

We still get warnings from specific usage of things like:

```
nix-repl> legacyPackages.aarch64-darwin.darwin.apple_sdk_11_0.callPackage
trace: evaluation warning: apple_sdk_11_0.callPackage: deprecated and will be removed in Nixpkgs 25.11; see <https://nixos.org/manual/nixpkgs/stable/#sec-darwin> for documentation and migration instructions
«lambda callPackageWith @ /nix/store/igkhqxk885qcxvzxqziqp9gjpszp9isd-source/lib/customisation.nix:240:15»
nix-repl> legacyPackages.aarch64-darwin.darwin.apple_sdk_12_3.objc4
trace: evaluation warning: apple_sdk_12_3.objc4: these stubs do nothing and will be removed in Nixpkgs 25.11; see <https://nixos.org/manual/nixpkgs/stable/#sec-darwin> for documentation and migration instructions. Add `apple-sdk_12` to build inputs instead to use the macOS 12 SDK.
«derivation /nix/store/ibsxq4b9jzf23388g83sasrnjjlp7mh7-objc4-12.3.drv»
```

though the wording is a little different in places and the references to `darwin` are sometimes missing.

@emilazy I would extremely greatly appreciate some reassurance that I didn't break anything here (and that the warnings do actually get emitted if they're relevant).

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

keysmashes committed:
darwin: reduce `nix search` warning spam

keysmashes committed:
darwin: lib.warn -> lib.warnOnInstantiate

emilazy commented on pkgs/os-specific/darwin/apple-sdk/mk-stub.nix:
```diff
@@ -1,14 +1,26 @@
+msg: version: pname:
+lib.warnOnInstantiate
+  (
+    if builtins.isString msg then
+      "${msg}.${pname}: ${common}"
+    else
+      "${msg.prefix}.${pname}: ${common} ${msg.suffix}"
```
This is just for “Add `apple-sdk_12` to build inputs instead to use the macOS 12 SDK.”, right? I think we can skip the complexity here: the migration instructions tell you how to use the new SDK pattern, and few people will be using the macOS 12 SDK with the old pattern to begin with anyway. (But you did already implement it, so it’s fine if you don’t want to touch it again.)

(The reason I used a different message that omits “these stubs do nothing” was because I was worried people would just remove them without reading further and still hope to get the macOS 12 SDK. But I don’t think it’s a big deal, if we are always prepending that prefix anyway.)

keysmashes replied:
Awesome, removing the `prefix`/`suffix` shenanigans makes a lot of the messiness go away, yeah.

emilazy commented on pkgs/os-specific/darwin/apple-sdk-11.0/default.nix:
```diff
@@ -249,17 +249,24 @@ lib.genAttrs [
+    (
+      name:
+      lib.warn
+        "apple_sdk_11_0.${name}: deprecated and will be removed in Nixpkgs 25.11; see <https://nixos.org/manual/nixpkgs/stable/#sec-darwin> for documentation and migration instructions"
```
```suggestion
        "darwin.apple_sdk_11_0.${name}: deprecated and will be removed in Nixpkgs 25.11; see <https://nixos.org/manual/nixpkgs/stable/#sec-darwin> for documentation and migration instructions"
```

Can you help me understand why these don’t cause warnings when `nix search` traverses into `llvmPackages_16` etc.?

keysmashes replied:
That is a great question. Looking at `nix search --help`, I _think_ this is because this attrset lacks a `recurseForDerivations = true`, so `nix search` doesn't actually evaluate its values?

emilazy replied:
Right, makes sense.

emilazy commented on pkgs/os-specific/darwin/apple-sdk-12.3/default.nix:
```diff
@@ -6,7 +6,10 @@
 let
-  mkStub = callPackage ../apple-sdk/mk-stub.nix { } "12.3";
+  mkStub = callPackage ../apple-sdk/mk-stub.nix { } {
+    prefix = "apple_sdk_12_3";
```
```suggestion
    prefix = "darwin.apple_sdk_12_3";
```

emilazy commented on pkgs/top-level/darwin-aliases.nix:
```diff
@@ -163,7 +154,8 @@ stubs
   ### S ###
 
   stdenvNoCF =
-    lib.warn "darwin.stdenvNoCF: use `stdenv` or `stdenvNoCC`; this will be removed in 25.11"
+    lib.warnOnInstantiate
+      "darwin.stdenvNoCF: use `stdenv` or `stdenvNoCC`; this will be removed in 25.11"
```
Correctly warns the first time `darwin.stdenvNoCF.mkDerivation` is used:

```
nix-repl> :b darwin.stdenvNoCF.mkDerivation { name = "test"; }
evaluation warning: darwin.stdenvNoCF: use `stdenv` or `stdenvNoCC`; this will be removed in 25.11
error: builder for '/nix/store/wn6zg4zh9as201sryrf4v15v3hra8j1v-test.drv' failed with exit code 1;
       last 2 log lines:
       > Running phase: unpackPhase
       > variable $src or $srcs should point to the source
       For full logs, run:
         nix log /nix/store/wn6zg4zh9as201sryrf4v15v3hra8j1v-test.drv
[0 built (1 failed)]
```

emilazy commented on pkgs/top-level/darwin-aliases.nix:
```diff
@@ -136,11 +127,11 @@ stubs
 
   libauto = throw "'darwin.libauto' has been removed, as it was broken and unmaintained"; # added 2024-05-10
-  libresolvHeaders = lib.warn "darwin.libresolvHeaders: use `lib.getInclude darwin.libresolv`; this will be removed in 25.11" (
+  libresolvHeaders = lib.warnOnInstantiate "darwin.libresolvHeaders: use `lib.getInclude darwin.libresolv`; this will be removed in 25.11" (
```
Seems to work:

```
nix-repl> "${darwin.libresolvHeaders}"
evaluation warning: darwin.libresolvHeaders: use `lib.getInclude darwin.libresolv`; this will be removed in 25.11
"/nix/store/q811742vhd5cfbz2chs0mnbxq6har5q2-libresolv-83-dev"

nix-repl> stdenv.mkDerivation { name = "test"; buildInputs = [ darwin.libresolvHeaders ]; }
evaluation warning: darwin.libresolvHeaders: use `lib.getInclude darwin.libresolv`; this will be removed in 25.11
evaluation warning: darwin.libresolvHeaders: use `lib.getInclude darwin.libresolv`; this will be removed in 25.11
evaluation warning: darwin.libresolvHeaders: use `lib.getInclude darwin.libresolv`; this will be removed in 25.11
«derivation /nix/store/wlxiznjqnpf5y207afz1yi2rlc6xhz23-test.drv»
```

I don’t know why it gets printed so many times in the latter case, but it probably doesn’t matter.

emilazy commented on pkgs/os-specific/darwin/apple-sdk-11.0/default.nix:
```diff
@@ -249,17 +249,24 @@ lib.genAttrs [
+//
+  lib.genAttrs
+    [
+      "callPackage"
```
Works:

```
nix-repl> darwin.apple_sdk_11_0.callPackage ({ }: 123) { }
evaluation warning: apple_sdk_11_0.callPackage: deprecated and will be removed in Nixpkgs 25.11; see <https://nixos.org/manual/nixpkgs/stable/#sec-darwin> for documentation and migration instructions
123
```

emilazy commented on pkgs/os-specific/darwin/apple-sdk-11.0/default.nix:
```diff
@@ -6,7 +6,7 @@
 
 let
-  mkStub = callPackage ../apple-sdk/mk-stub.nix { } "11.0";
+  mkStub = callPackage ../apple-sdk/mk-stub.nix { } "apple_sdk_11_0" "11.0";
```
```suggestion
  mkStub = callPackage ../apple-sdk/mk-stub.nix { } "darwin.apple_sdk_11_0" "11.0";
```

No reason not to keep the `darwin.` breadcrumb here, right? It’s the canonical attribute path.

emilazy reviewed (commented):
Thanks, this generally seems to work great!

keysmashes commented:
Thank you very much for the testing! I've addressed your comments I think (added `darwin.` in a couple of places and dropped the note about `apple-sdk_12`).

emilazy reviewed (approved):
Looks great, thank you. I will try to be diligent about using `lib.warnOnInstantiate` going forward. FWIW, the warning for CUDA will be harder to deal with since there’s all sorts of stuff in that attribute set; I guess we might need some kind of automated thing that adds `lib.warnOnInstantiate` to derivations and `lib.warn` to everything else.
